### PR TITLE
[Yaml] Added support for parsing PHP constants

### PR DESCRIPTION
**Pull request by HeahDude**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | TODO |

### CODE REVIEW

**Comment by HeahDude** [comment]
I was reading http://symfony.com/doc/current/components/dependency_injection/parameters.html#constants-as-parameters and felt sad about the note for yaml support.

Is this PR acceptable? Does this need a flag? Or to be reversible (which does not seem doable)?

There is no sense to handle such case with the dumper imho (unless it is passed as a string?? but what would be the use case?).

**Comment by hhamon** [comment]
I'm -1 for such horrible hacks. YAML is not a suitable format to handle metada like describing a thing is a constant.

**Comment by xabbuh** [comment]
Usually I would agree with @hhamon and I had the same feeling when reading the headline of this PR. But on the other hand @fabpot always wanted to have the Yaml component to support the things we need for Symfony (i.e. being able to configure the framework) instead of supporting the full YAML specs. Now, this is something that is supported by other configuration formats (i.e. PHP and XML), but which isn't supported for the YAML config format. Thus, I am :+1: for closing the gap.

File: src/Symfony/Component/Yaml/Inline.php [hunk 1]
```diff
@@ -562,6 +562,8 @@ private static function evaluateScalar($scalar, $flags, $references = array())
                         }
 
                         return;
+                    case 0 === strpos($scalar, '!php/const:'):
+                        return @constant(substr($scalar, 11));
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: I think we should to throw a `ParseException` here if the constant does not exist (or at least do that when the `PARSE_EXCEPTION_ON_INVALID_TYPE` flag is set).

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: I agree, I first did it because this flag looks indeed appropriate. I don't know why I revert it.

So I'll push it quickly :)

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: Well, now that I think of this again, I do not think we should evaluate the flag for this. Such an error imo should never be silenced.

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: It's not an error, it's a warning, preventing to get `null`, remove the `@` and the last test fail because it gets `0.0` instead of `null`. That's why I silenced it (and removed the flag in the process without giving it a second thought).

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: What I mean is that this usually indicates a mistake made by the developer when writing the YAML file.

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: Hm, on the other hand they may have to deal with YAML files that did not write themself. But that's rather and argument to completely turn off parsing constants by default. Thus having the developer deliberately turn it on when wanting to parse constants.

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: I guess getting `null` should be expected (a constant could be defined in an environment and not in another), the flag let the user decide it in a known and explicit way.

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: Ok I now see you answered before me, I maintain my proposal though, but I'm not against yours.
 
Still I would rather not add a new flag for this, the goal here is to ease configuration ; currently you need to import an `xml` or a `php` file while `yaml` is very common (I think and always used it only).

So needing to add a flag does not seem to "keep it easy" while using `PARSE_EXCEPTION_ON_INVALID_TYPE` looks more like "keep doing as usual, just pass a constant if you want".

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: @HeahDude Well, for the configuration of services the users would not have to do that. It's the responsibility of us to set up the parser properly in the `YamlFileLoader`. So we are more talking about someone who uses the parser in their own code.

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: Understood! So I need to:

1. add a flag like `PARSE_PHP_CONSTANTS` in `Yaml` class

2. either throw a `ParseException` or return `null` when the constant does not exists
   maybe using `PARSE_EXCEPTION_ON_INVALID_TYPE`

3. add the flag `PARSE_PHP_CONSTANTS` to `YamlFileLoader` config.

Is that correct? What about point 2?

**Comment by iltar** [comment]
I would be really happy to see this implemented. I prefer yaml over xml by a lot for non-public packages, but not having constants is a big limitation to yaml :(

File: src/Symfony/Component/Yaml/Inline.php (hunk 1 above)
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: I agree. And I think as you would then have to deliberately turn on the feature, throwing an exception when a constant does not exist is the right way.

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: I'm almost there :)

Since the constant are not autoloaded (IIUC), one may expect to get `null` if a constant is not defined in an environment, so I guess we can take advantage of the triggered `E_WARNING` (ref http://php.net/manual/en/function.constant.php) as it will be shown as a silenced error in the profiler thanks to https://github.com/symfony/symfony/pull/16760.

Hence, I suggest we only trigger an exception when `PARSE_EXCEPTION_ON_INVALID_TYPE` flag is passed.

I already did such implementation but I've got an issue when the constant is not "inlined":
```yaml
foo:
    - '!php/const:PHP_INT_MAX' # does not work
bar: ['!php/const:PHP_INT_MAX'] # works
baz: { foo: '!php/const:PHP_INT_MAX' } # works
```

I think it comes from the `Parser` as I only changed the `Inline` class for now, any lead ?

Thanks.

**Comment by HeahDude** [comment]
Changelog:

 - f23dd9b addressed comments in original commit

 - 56297da added support in the DI `YamlFileLoader`

 - a58d960 added failing test cases

I was wrong in my previous comment it's also failing because the parser ignores special values when they are "nested", i.e: `foo: '!php/const:PHP_INT_MAX'`

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+3.2.0
+-----
+
+ * added support for PHP constants in yaml configuration files
```
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/CHANGELOG.md]: YAML

File: src/Symfony/Component/DependencyInjection/composer.json [hunk 3]
```diff
@@ -19,7 +19,7 @@
         "php": ">=5.5.9"
     },
     "require-dev": {
-        "symfony/yaml": "~2.8|~3.0",
+        "symfony/yaml": "~2.8|~3.2",
```
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/composer.json]: Tests will fail on 2.8 now.

File: src/Symfony/Component/Yaml/Inline.php [hunk 4]
```diff
@@ -562,6 +564,18 @@ private static function evaluateScalar($scalar, $flags, $references = array())
...
+                    case 0 === strpos($scalar, '!php/const:'):
+                        if (self::$constantSupport) {
+                            $constant = @constant(substr($scalar, 11));
+
+                            if (null === $constant && self::$exceptionOnInvalidType) {
+                                throw new ParseException(sprintf('The constant "%s" is not defined.', substr($scalar, 11)));
+                            }
+
+                            return $constant;
+                        }
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: This must throw an exception when constant support is disabled and `$exceptionOnInvalidType` is set (like we do for object support).

File: src/Symfony/Component/Yaml/Inline.php [hunk 5]
```diff
@@ -562,6 +564,18 @@ private static function evaluateScalar($scalar, $flags, $references = array())
                         }
 
                         return;
+                    case 0 === strpos($scalar, '!php/const:'):
+                        if (self::$constantSupport) {
+                            $constant = @constant(substr($scalar, 11));
+
+                            if (null === $constant && self::$exceptionOnInvalidType) {
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: This will fail the case where the constant was explicitly set to `null`.

File: src/Symfony/Component/DependencyInjection/composer.json (hunk 3 above)
Comment (HeahDude) [inline src/Symfony/Component/DependencyInjection/composer.json]: Noted

File: src/Symfony/Component/Yaml/Inline.php (hunk 5 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: Arf! So the `E_WARNING` should be enough then, right?

File: src/Symfony/Component/Yaml/Inline.php (hunk 4 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: Alright!

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 6]
```diff
@@ -1201,6 +1201,33 @@ public function getInvalidBinaryData()
...
+
+        // Returns a string by default
+        $this->assertSame($yaml, $this->parser->parse($yaml));
+        // Returns a constant value when passing flag
+        $this->assertSame(PHP_INT_MAX, $this->parser->parse($yaml, Yaml::PARSE_CONSTANT));
+    }
+
+    public function testParseNestedPhpConstant()
+    {
+        $yaml = "foo: '!php/const:PHP_INT_MAX'";
```
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: What about this failing test?

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: The tag will never be evaluated as you quoted the value (this is then just a simple string in YAML).

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: Thanks, I've already understood my mistake and updated the tests, will push it as soon as I can :)

**Comment by HeahDude** [comment]
Ok, comments addressed, tests are green :)

I guess the failure on low depth will have to wait for master to point on 3.2.

**Comment by dunglas** [comment]
Supporting PHP constants would be nice but the proposed syntax `!php/const:PHP_INT_MAX` is not intuitive and looks ugly. Can we have another nicer syntax?

**Comment by HeahDude** [comment]
@dunglas suggestions are welcome! It actually matches the object syntax `!php/object`, so I thought it would be both consistent and easy to learn. Have you something else in mind?

**Comment by HeahDude** [comment]
Could be like float or binary `!! constant PHP_INT_MAX`.

File: src/Symfony/Component/Yaml/Inline.php [hunk 7]
```diff
@@ -562,6 +564,15 @@ private static function evaluateScalar($scalar, $flags, $references = array())
                         }
 
                         return;
+                    case 0 === strpos($scalar, '!php/const:'):
+                        if (self::$constantSupport) {
+                            return @constant(substr($scalar, 11));
```
Comment (stof) [inline src/Symfony/Component/Yaml/Inline.php]: receiving null when a constant does not exist makes debugging very hard when you make a typo

**Comment by stof** [comment]
> Could be like float or binary !! constant PHP_INT_MAX.

no it cannot. ``!!float`` is part of the Yaml standard. custom tags must star with a single ``!``, not 2

File: src/Symfony/Component/Yaml/Inline.php (hunk 7 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: @stof there is a hidden conversation on a diff above between me and @xabbuh where we argue about this. It returns `null` but there is a warning logged as silenced error thanks to #16760.
Isn't that enough? Maybe we could also silence it in master for the `XmlLoader`?

Comment (stof) [inline src/Symfony/Component/Yaml/Inline.php]: well, IMO, we should not silence things here

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: It would be consistent with the current `XmlLoader` indeed. But `null` will be return anyway with an `E_WARNING`.

**Comment by HeahDude** [comment]
Ok, so why not `!constant PHP_INT_MAX`? Although I'm not against the current syntax.

**Comment by xabbuh** [comment]
I would not change the tag name. The `php:` part acts like a Namespace here which reduces the risk to conflict with custom tags from other libraries.

**Comment by dunglas** [comment]
@xabbuh it makes perfectly sense but `!constant PHP_INT_MAX` is really nicer.

**Comment by fabpot** [comment]
I'm against adding something that does not follow the YAML specification.

**Comment by dunglas** [comment]
If I understand correctly the spec, both examples are valid: http://www.yaml.org/spec/1.2/spec.html#id2782728

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 8]
```diff
@@ -22,6 +22,7 @@
 use Symfony\Component\Yaml\Exception\ParseException;
 use Symfony\Component\Yaml\Parser as YamlParser;
 use Symfony\Component\ExpressionLanguage\Expression;
+use Symfony\Component\Yaml\Yaml;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: can you please move it one line up in the yaml group?

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 9]
```diff
@@ -366,7 +367,7 @@ protected function loadFile($file)
         }
 
         try {
-            $configuration = $this->yamlParser->parse(file_get_contents($file));
+            $configuration = $this->yamlParser->parse(file_get_contents($file), Yaml::PARSE_CONSTANT);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: this means that the DI component not can't work with Yaml < 3.2, isn't it? So the composer.json file should get a conflict rule. Or: would it make sense to enable const parsing by default (and thus not require any change here?)

File: src/Symfony/Component/Yaml/Inline.php (hunk 7 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: We shouldn't we trigger an exception when the const is not defined? Would look better to me ATM

File: src/Symfony/Component/Yaml/Inline.php [hunk 10]
```diff
@@ -28,6 +28,7 @@ class Inline
     private static $exceptionOnInvalidType = false;
     private static $objectSupport = false;
     private static $objectForMap = false;
+    private static $constantSupport = false;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: Are we sure we need a flag here? Couldn't we unconditionally enable this?

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 8 above)
Comment (HeahDude) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: Are you sure about this? The "good way" would be to move up the `Expression` line but as a CS fix it shouldn't be done in master, right?

File: src/Symfony/Component/DependencyInjection/composer.json [hunk 11]
```diff
@@ -19,7 +19,7 @@
         "php": ">=5.5.9"
     },
     "require-dev": {
-        "symfony/yaml": "~2.8|~3.0",
+        "symfony/yaml": "~3.2",
```
Comment (HeahDude) [inline src/Symfony/Component/DependencyInjection/composer.json]: @nicolas-grekas

>this means that the DI component not can't work with Yaml < 3.2, isn't it?

Isn't this line enough?

File: src/Symfony/Component/Yaml/Inline.php (hunk 10 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: @xabbuh what do you think?

File: src/Symfony/Component/Yaml/Inline.php (hunk 7 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: Is there a way to distinguish `null` when assigned to a constant from when it's not defined?

Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: yes, using `defined()`

File: src/Symfony/Component/DependencyInjection/composer.json (hunk 11 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/composer.json]: this line does apply only when testing the component, it has no effect when used as a dep of something else

File: src/Symfony/Component/Yaml/Inline.php (hunk 7 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: OK great! Then we should do it indeed. What do others think? @xabbuh @stof

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 8 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: let's group yaml lines now that we can (but we don't move existing lines to reduce merge conflicts)

File: src/Symfony/Component/Yaml/Inline.php (hunk 10 above)
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: Depends on whether we want to throw exceptions when a constant is not defined. Imo the flag makes sense if we then throw an exception when a constant does not exist (which would make debugging a lot easier than when we just returned `null`). Though if we enabled this feature unconditionally, we would not be able to throw exceptions when the constant was not defined.

Comment (stof) [inline src/Symfony/Component/Yaml/Inline.php]: Throwing the exception already depends on the ``exceptionOnInvalidType`` flag currently

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: Ok then, they are many options.

1. Parse constants by defaults:
 ```php
case 0 === strpos($scalar, '!php/const:'):
    if (defined($const = substr($scalar, 11))) {
        return constant($const);
    }
    if (self::$exceptionOnInvalidType) {
        throw new ParseException(sprintf('The PHP constant "%s" is not defined.', $const));
    }

    return;
 ```

2. Or using a flag as is:
 ```php
// use Symfony\Component\Yaml\Exception\RuntimeException;

case 0 === strpos($scalar, '!php/const:'):
    if (self::$constantSupport) {
        $const = substr($scalar, 11);
        if (defined($const)) {
            return constant($const);
        }

        throw new RuntimeException(sprintf('The constant "%s" is not defined.', $const));
    }
    if (self::$exceptionOnInvalidType) {
        throw new ParseException(sprintf('The string "%s" could not be parsed as constant. Have you forgotten to use "Yaml::PARSE_CONSTANT" flag?', $scalar));
    }

    return;
 ```

3. Suggestions?

For now I vote for option 2, since php objects are not parsed by default.

Comment (javiereguiluz) [inline src/Symfony/Component/Yaml/Inline.php]: My vote is for a flag too because this component should be as close to the Yaml spec as possible. `!php/const:` is a Symfony thing, not a Yaml thing.

**Comment by fabpot** [comment]
@HeahDude Can you finish this PR?

File: src/Symfony/Component/Yaml/Inline.php (hunk 10 above)
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: I agree with that. It will consistent with how we already treat object PHP support.

**Comment by HeahDude** [comment]
Yes! With option 2 from my comment above?

**Comment by HeahDude** [comment]
Comments addressed. Waiting for tests to be green, needs review, thanks!

**Comment by fabpot** [comment]
👍

File: src/Symfony/Component/Yaml/Inline.php (hunk 10 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: Fine by me, then [my comment above](https://github.com/symfony/symfony/pull/18626/files#r66247758) applies.

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 9 above)
Comment (HeahDude) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: Fixed

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 12]
```diff
@@ -366,7 +367,7 @@ protected function loadFile($file)
...
         } catch (ParseException $e) {
```
Comment (HeahDude) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: The `Symfony\Component\Yaml\Exception\RuntimeException` might not be caught here, @xabbuh any ideas on how we should handle it?

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 9 above)
Comment (HeahDude) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: @nicolas-grekas see https://github.com/symfony/symfony/pull/18626/files#diff-85ef95bd09276e850971f82e010b284dR32

File: src/Symfony/Component/Yaml/Inline.php [hunk 13]
```diff
@@ -581,6 +584,19 @@ private static function evaluateScalar($scalar, $flags, $references = array())
                         }
 
                         return;
+                    case 0 === strpos($scalar, '!php/const:'):
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Inline.php]: What about:
```php
                    case 0 === strpos($scalar, '!php/const:'):
                        if (self::$constantSupport && defined($const = substr($scalar, 11))) {
                            return constant($const);
                        }
                        if (self::$exceptionOnInvalidType) {
                            throw new ParseException(sprintf('The "%s" YAML string could not be parsed because the %s.', $scalar, self::$constantSupport ? ' "Yaml::PARSE_CONSTANT" flag was not set' : 'corresponding PHP constant was not defined'));
                        }

                        return; 
```

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 12 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: @Haehnchen see https://github.com/symfony/symfony/pull/18626/files#r67163458

File: src/Symfony/Component/Yaml/Inline.php (hunk 13 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: @nicolas-grekas 👍

**Comment by nicolas-grekas** [comment]
Status: needs work

File: src/Symfony/Component/Yaml/Inline.php (hunk 13 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: @nicolas-grekas Working on it I think we should throw an exception when the constant is not defined even if `self::$exceptionOnInvalidType` is false, what do you think **@symfony/deciders**?

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Inline.php]: In that case maybe I should only change the YAML `RuntimeException` to a `ParseException` so it's caught in the DI loader.

Comment (fabpot) [inline src/Symfony/Component/Yaml/Inline.php]: I think throwing an exception when a constant is not defined is a good idea.

**Comment by HeahDude** [comment]
Updated. Status: Needs Review

**Comment by HeahDude** [comment]
Status: Needs Review

**Comment by fabpot** [comment]
👍 (failures not related)

File: src/Symfony/Component/Yaml/Inline.php [hunk 14]
```diff
@@ -581,6 +583,19 @@ private static function evaluateScalar($scalar, $flags, $references = array())
...
+                        if (self::$constantSupport) {
+                            if (defined($const = substr($scalar, 11))) {
+                                return constant($const);
+                            }
+
+                            throw new ParseException(sprintf('The constant "%s" is not defined.', $const));
+                        }
+                        if (self::$exceptionOnInvalidType) {
+                            throw new ParseException(sprintf('The string "%s" could not be parsed as constant. Have you forgotten to pass "Yaml::PARSE_CONSTANT" flag to the parser?', $scalar));
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: [...] pass **the** [...]

File: src/Symfony/Component/Yaml/Tests/InlineTest.php [hunk 15]
```diff
@@ -36,6 +36,34 @@ public function testParseWithMapObjects($yaml, $value)
...
+        $actual = Inline::parse($yaml, Yaml::PARSE_CONSTANT);
+
+        $this->assertSame($value, $actual);
+    }
+
+    public function testParsePhpConstantThrowsExceptionWhenUndefined()
+    {
+        $expectedMessage = 'The constant "WRONG_CONSTANT" is not defined';
+
+        $this->setExpectedException('\Symfony\Component\Yaml\Exception\ParseException', $expectedMessage);
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/InlineTest.php]: Can you use the `@expectedException` annotation instead?

File: src/Symfony/Component/Yaml/Tests/InlineTest.php [hunk 16]
```diff
@@ -36,6 +36,34 @@ public function testParseWithMapObjects($yaml, $value)
...
+
+        Inline::parse('!php/const:WRONG_CONSTANT', Yaml::PARSE_CONSTANT);
+    }
+
+    public function testParsePhpConstantThrowsExceptionOnInvalidType()
+    {
+        $expectedMessage = '#The string "!php/const:PHP_INT_MAX" could not be parsed as constant.*#';
+
+        $this->setExpectedExceptionRegExp('\Symfony\Component\Yaml\Exception\ParseException', $expectedMessage);
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/InlineTest.php]: same here

File: src/Symfony/Component/Yaml/Inline.php (hunk 14 above)
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: [...] be parsed as **a** constant.

File: src/Symfony/Component/Yaml/Tests/InlineTest.php [hunk 17]
```diff
@@ -442,6 +470,16 @@ public function getTestsForParseWithMapObjects()
...
 
+    public function getTestsForParsePhpConstants()
+    {
+        return array(
+            array('!php/const:Symfony\Component\Yaml\Yaml::PARSE_CONSTANT', Yaml::PARSE_CONSTANT),
+            array('!php/const:PHP_INT_MAX', PHP_INT_MAX),
+            array('[!php/const:PHP_INT_MAX]', array(PHP_INT_MAX)),
+            array('{ foo: !php/const:PHP_INT_MAX }', array('foo' => PHP_INT_MAX)),
+        );
+    }
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/InlineTest.php]: I would move the data provider just below the `testParsePhpConstants` test.

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 18]
```diff
@@ -1351,6 +1351,30 @@ public function parserThrowsExceptionWithCorrectLineNumberProvider()
...
+        $this->parser->parse('foo: !php/const:WRONG_CONSTANT', Yaml::PARSE_CONSTANT);
+    }
+
+    public function testParsePhpConstantThrowsExceptionOnInvalidType()
+    {
+        $expectedMessage = '#The string "!php/const:PHP_INT_MAX" could not be parsed as constant.*#';
+        $this->setExpectedExceptionRegExp('\Symfony\Component\Yaml\Exception\ParseException', $expectedMessage);
+
+        $this->parser->parse('foo: !php/const:PHP_INT_MAX', Yaml::PARSE_EXCEPTION_ON_INVALID_TYPE);
+    }
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: Not sure if it's really necessary to duplicate all the tests from the `Inline` class here.

**Comment by fabpot** [comment]
@HeahDude Can you take @xabbuh comments into account?

File: src/Symfony/Component/Yaml/Tests/ParserTest.php (hunk 18 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: @xabbuh not "all" so we'd just keep the one without exception?

Comment (HeahDude) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: @xabbuh or do you mean I can remove all of them?

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: I would simply revert the changes in this file.

File: src/Symfony/Component/Yaml/Tests/InlineTest.php (hunk 17 above)
Comment (HeahDude) [inline src/Symfony/Component/Yaml/Tests/InlineTest.php]: Done

**Comment by fabpot** [comment]
Thank you @HeahDude.

**Comment by teohhanhui** [comment]
It's unfortunate that we did not consider it important to adhere to the YAML spec: http://www.yaml.org/spec/1.2/spec.html#tag/local/
